components/suggest/src/yelp.rs

```diff
@@ -0,0 +1,319 @@
+            named_params! {
+                ":keyword": sign.keyword,
+                ":need_location": sign.need_location,
+            },
+        )?;
+    }
+
+    scope.err_if_interrupted()?;
+    conn.execute(
+        "INSERT INTO yelp_templates(url) VALUES(?)",
+        [&dataset.template],
+    )?;
+
+    Ok(())
+}
+
+/// Fetch Yelp suggestion from given user's query.
+pub(crate) fn fetch_suggestion(
+    conn: &Connection,
+    query: &SuggestionQuery,
+) -> Result<Option<Suggestion>> {
+    if !query.providers.contains(&SuggestionProvider::Yelp) {
+        return Ok(None);
+    }
+
+    if query.keyword.len() > MAX_QUERY_LENGTH {
+        return Ok(None);
+    }
+
+    if !&query.keyword.contains(" ") {
+        // Try to create by one word.
+        return to_suggestion(conn, query, &query.keyword, None, None, None, None, false);
+    }
+
+    // Find the location sign and the location.
+    let (query_without_location, location_sign, location, need_location) =
+        find_location(conn, &query.keyword)?;
+
+    if let (Some(_), false) = (&location, need_location) {
+        // The location sign does not need the specific location, but user is setting something.
+        return Ok(None);
+    }
+
+    if query_without_location.is_empty() {
+        // No remained query.
+        return Ok(None);
+    }
+
+    // Find the modifiers.
+    let (subject_candidate, pre_modifier, post_modifier) =
+        find_modifiers(conn, &query_without_location)?;
+
+    to_suggestion(
+        conn,
+        query,
+        &subject_candidate,
+        pre_modifier,
+        post_modifier,
+        location_sign,
+        location,
+        need_location,
+    )
+}
+
+fn to_suggestion(
+    conn: &Connection,
+    query: &SuggestionQuery,
+    subject: &str,
+    pre_modifier: Option<String>,
+    post_modifier: Option<String>,
+    location_sign: Option<String>,
+    location: Option<String>,
+    need_location: bool,
+) -> Result<Option<Suggestion>> {
+    if !is_subject(conn, subject)? {
+        // No subject.
+        return Ok(None);
+    }
+
+    // Build URL.
+    let mut description = Vec::<&str>::new();
+
+    if let Some(modifier) = &pre_modifier {
+        description.append(&mut modifier.split_whitespace().collect::<Vec<&str>>());
+    }
+
+    description.append(&mut subject.split_whitespace().collect::<Vec<&str>>());
+
+    if let Some(modifier) = &post_modifier {
+        description.append(&mut modifier.split_whitespace().collect::<Vec<&str>>());
+    }
+
+    if let (Some(sign), false) = (&location_sign, need_location) {
+        // This location sign such the 'near by' needs to add as a description parameter.
+        description.append(&mut sign.split_whitespace().collect::<Vec<&str>>());
+    }
+
+    let mut url_template: String = conn
+        .query_row_and_then("SELECT url FROM yelp_templates", [], |row| {
+            row.get::<_, String>(0)
+        })?
+        .replace("{description}", &description.join("+"));
```

Comment:
To summarize, you're splitting the modifiers and subject into vecs of words, putting them all in `description`, and then joining the words with `"+"`, right? And the reason is to URL-encode the description, essentially, right? I think we might want to do some proper encoding here, which will be safer in case the subject or modifiers contain characters that should be encoded. I did a quick search for a Rust equivalent to `encodeURIComponent()` and found [this](https://rust-lang-nursery.github.io/rust-cookbook/encoding.html), and then did a quick search in the application-services repo and found that the `places` component [already uses](https://github.com/mozilla/application-services/blob/1d259497eb43e6fb416667dbdf81723e9e04d2e2/components/places/src/match_impl.rs#L371-L374) the [percent_encoding](https://docs.rs/percent-encoding/2.3.1/percent_encoding/) crate. So we can use that here too, and then we don't have to worry about vetting it when we vendor this code into m-c. Does that sound OK?

Reply:
Thanks! Yes, I intended to do that.  And I did the `percent-encoding` using `Url::parse()`.

The reason why I have done this is if we search on Yelp, it seems that they replace the whitespace of keywords with `+'` not `%20`.
e.g. https://www.yelp.com/search?find_desc=Ramen+Near+Me&find_loc=San+Francisco
So, to make it exactly match with the Yelp URL, I did it. I will try to find a better way.

Reply:
https://docs.rs/form_urlencoded/latest/form_urlencoded/ ?

Reply:
> and then did a quick search in the application-services repo and found that the `places` component [already uses](https://github.com/mozilla/application-services/blob/1d259497eb43e6fb416667dbdf81723e9e04d2e2/components/places/src/match_impl.rs#L371-L374) the [percent_encoding](https://docs.rs/percent-encoding/2.3.1/percent_encoding/) crate. So we can use that here too, and then we don't have to worry about vetting it when we vendor this code into m-c.

Correction w/r/t vetting, this probably isn't true because `places` isn't used in m-c, but I'm not 100% sure.

Reply:
It looks like `percent_encoding` and `form_urlencoded` are both already vendored and vetted (🎉), because they're dependencies of `url`.

[Bug 1866135](https://bugzilla.mozilla.org/show_bug.cgi?id=1866135) was the last one to bump the vendored and vetted versions for both  `percent_encoding` and `form_urlencoded`, so I think either one should work, without extra work from us!